collection.go

```diff
@@ -594,39 +604,40 @@ func assignValues(to interface{}, valueOf func(reflect.Type, string) (reflect.Va
 		return fmt.Errorf("nil pointer to %T", to)
 	}
 
-	if err := flattenStruct(toValue.Elem()); err != nil {
-		return err
+	// Follow the pointer to the provided object and collect
+	// its fields that were tagged with 'ebpf'.
+	fields, err := ebpfFields(toValue.Elem(), nil)
+	if err != nil {
+		return fmt.Errorf("gathering field tags: %w", err)
 	}
 
-	type elem struct {
-		// Either *Map or *Program
-		typ  reflect.Type
-		name string
-	}
+	// Store assigned objects and the struct fields they are assigned to.
+	assigned := make(map[interface{}]structField)
 
-	assignedTo := make(map[elem]string)
 	for _, field := range fields {
-		name := field.Tag.Get("ebpf")
-		if strings.Contains(name, ",") {
+		// Get string value the field is tagged with.
+		tag := field.Tag.Get("ebpf")
+		if strings.Contains(tag, ",") {
 			return fmt.Errorf("field %s: ebpf tag contains a comma", field.Name)
 		}
 
-		e := elem{field.Type, name}
-		if assignedField := assignedTo[e]; assignedField != "" {
-			return fmt.Errorf("field %s: %q was already assigned to %s", field.Name, name, assignedField)
-		}
-
-		value, err := valueOf(field.Type, name)
+		// Get the eBPF object referred to by the tag.
+		value, err := getValue(field.Type, tag)
 		if err != nil {
 			return fmt.Errorf("field %s: %w", field.Name, err)
 		}
 
+		// Use the interface value as a unique key, since they're comparable.
+		if af, ok := assigned[value]; ok {
```

Comment:
I see why this works, but it makes the code kind of subtle. e.g. we now require that getValue always returns the same value for a given type, tag argument for this check to succeed. I think I prefer the explicit `{reflect.Type, name}` check.

Reply:
After offline discussion, since `assigned` doesn't technically need to be exported, this can be reverted.